### PR TITLE
Remove default partition for equality filters

### PR DESCRIPTION
**Pull request by THANATOSLAVA**
Background: In static partition selection, if the query has non-zero general filters, all the filters are treated as general filters. This triggers default partition scan for levels with default partition, even if the filter of that level is an equality filter. This potentially elongates the query runtime.

Solution: Add all equality filters to the equality filter list.

Implementation: 1. In non-passthrough filter translation (Expr -> DXL), add equality filter to the equality filter list. Add non-NULL general predicates containing disjunctions of ident equal comparisons to the equality filter list. 2. Rewrite filter list translation (DXL -> Plstmt) by handling equality filter elem list in DXL -> Scalar translation. This makes use of the recursive nature of DXL -> Scalar translation, and improves code readability. 3. Rewrite partition selection for equality prediçates. This allows the removal of the equality filter elem list wrapper when the list only has one child (const, cast, ident...) 4. Update filter lists in MDP tests. Relax file name matching in fix mdp script. 5. Add test for ORCA avoiding default partition scan for all levels with equality predicates.

### CODE REVIEW

**Comment by THANATOSLAVA** [comment]
MDP test passed

**Comment by dgkimura** [comment]
Nit - to help myself (and other reviewers), could you break your single commit into 5 commits for each change you listed in the description?  I think that would help make it easier to review each logical change in this PR.  Otherwise, I find this a little hard to follow and review.

**Comment by THANATOSLAVA** [comment]
> 

Thanks David for helping rewriting the commits~

File: src/test/regress/expected/partition_pruning.out [hunk 1]
```diff
@@ -3466,4 +3466,314 @@ on (date_trunc('day', r.pkey) = l.pkey
...
+-- Level 0: year 2 parts
+-- Level 1: quarter 5 parts
+-- Level 2: region 3 parts
+-- Level 3: sales 4 parts
+-- Total partitions: 2*5*3*4 = 120
+DROP TABLE sales;
+CREATE TABLE sales (id bigint, year int, quarter int, region text, sales int) DISTRIBUTED BY (id)
+PARTITION BY RANGE(year)
+ SUBPARTITION BY LIST (quarter)
+ SUBPARTITION TEMPLATE (DEFAULT subpartition other_quarter,
```
Comment (chrishajas) [inline src/test/regress/expected/partition_pruning.out:3479]: We might be able to make this test case a bit smaller (fewer subpartitions) to be more specific about what we're testing

**Comment by chrishajas** [comment]
Nice job! The code change looks good to me and thanks for adding some tests!. I am still a bit concerned about unknowns since there isn't much previous test coverage here unfortunately. We don't have many tests with multi-level partitions with default partitions, so I think adding and thinking of additional cases here that we can also add to the master branch would be helpful and give us more confidence.

For other reviewers, this builds on the work done in 6d041943f5bfcd303d2e598ab4e389f8741ee514 which may be helpful for additional context.

File: src/backend/gporca/data/dxl/minidump/BitmapIndexProbeMergeFilters.mdp [hunk 2]
```diff
@@ -728,9 +728,7 @@ EXPLAIN SELECT * FROM table_with_two_indices_ao WHERE b = 15 AND c BETWEEN 10000
             </dxl:Properties>
             <dxl:ProjList/>
             <dxl:PartEqFilters>
-              <dxl:PartEqFilterElems>
-                <dxl:ConstValue TypeMdid="0.23.1.0" Value="15"/>
-              </dxl:PartEqFilterElems>
```
[the remark this answers is not in the document]
Comment (THANATOSLAVA) [reply src/backend/gporca/data/dxl/minidump/BitmapIndexProbeMergeFilters.mdp:733]: Equality filter started off having just one nested node. The filter element list was added to support filters like `WHERE attr = a OR attr = b`, where both equality filters are appended to the element list. Ever since, even single elements were added to a list to work with the updated partitions selector, which only accepted lists. The code update basically expanded the support to both lists and elements.

File: src/backend/gporca/scripts/fix_mdps.py [hunk 3]
```diff
@@ -60,6 +60,8 @@ def processLogFile(logFileLines):
 
     for line in logFileLines:
         fileNameMatch = re.match(r'.*../data/dxl/minidump/*.', line)
+        if not fileNameMatch:
+		fileNameMatch = re.match(r'.*../data/dxl/multilevel-partitioning/*.', line)
```
[the remark this answers is not in the document]
Comment (THANATOSLAVA) [reply src/backend/gporca/scripts/fix_mdps.py:64]: Replaced tab with 4 spaces

File: src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp [hunk 4]
```diff
@@ -4875,18 +4875,14 @@ CTranslatorExprToDXL::PdxlPartEqFilterList(CExpression *pexpr)
 		CDXLNode *eqfilter_dxlnode = NULL;
 
 		if (NULL == pexprEqFilter && NULL == pexprFilter)
-		{
 			eqfilter_dxlnode = CTranslatorExprToDXLUtils::PdxlnBoolConst(
 				m_mp, m_pmda, true /*value*/);
-		}
```
[the remark this answers is not in the document]
Comment (THANATOSLAVA) [reply src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp:4881]: Oh do we enforce curly braces to guard single-line if statements in cpp files? If so, I'd be happy to add that. I personally find it a little less aesthetic.

[the remark this answers is not in the document]
Comment (dgkimura) [reply src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp:4881]: If we don't enforce that yet, then I think we should.  In master branch I think we can just add "readability-braces-around-statements" check to `.clang-tidy`.

File: src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp [hunk 5]
```diff
@@ -5328,41 +5316,34 @@ CTranslatorExprToDXL::ConstructLevelFilters4PartitionSelector(
...
-					CTranslatorExprToDXLUtils::PdxlnListFilterScCmp(
-						m_mp, m_pmda, pdxlnPartKey, pdxlnEq, pmdidTypePartKey,
-						pmdidTypeOther, IMDType::EcmptEq, ulLevel,
-						fDefaultPartition);
-			}
+			(*ppdxlnEqFilters)->AddChild(PdxlnScalar(pexprEqFilter));
+			(*ppdxlnFilters)
+				->AddChild(CTranslatorExprToDXLUtils::PdxlnBoolConst(
+					m_mp, m_pmda, true /*value*/));
+			continue;
```
[the remark this answers is not in the document]
Comment (THANATOSLAVA) [reply src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp:5328]: Good point. The existing design says, the filters are either "passed through" to the equality filter list (all equality filters), or, all added to the general filter list (all treated as general filters). With the code update, it's more like, the filters are either "passed through" (special case), or, depending on the situation, they may be added to equality or general filter list (general case). I kept the `continue` cause I find it more readable than `if`, `else if`. I added more comments to separate the three scenarios that we deal with in the for loop.

File: src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (THANATOSLAVA) [reply src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp:4881]: Sounds good.

File: src/test/regress/expected/partition_pruning.out (hunk 1 above)
Comment (THANATOSLAVA) [reply src/test/regress/expected/partition_pruning.out:3479]: The additional hierarchy was meant to persuade myself that I didn't break anything while meddling with sub/partition by range/list of datatype int/text using/not using template. I couldn't explain why the original code handled the equality filters differently depending on the partitions by range vs. list. But seems like removing that complexity nothing happened...

File: src/test/regress/sql/partition_pruning.sql [hunk 6]
```diff
@@ -936,4 +936,51 @@ on (date_trunc('day', r.pkey) = l.pkey
...
+quarter IN (2, 4) AND
+region = 'usa' AND
+sales = 8000;
+
+-- year 1, quarter 3 (including default), region 1, sales 3 (including default)
+-- Expect to scan partitions 1*3*1*3 = 9
+EXPLAIN (COSTS OFF) SELECT * FROM sales WHERE year > 2009 AND
+quarter >= 3 AND
+region = 'usa' AND
+sales > 8000;
```
[the remark this answers is not in the document]
Comment (THANATOSLAVA) [reply src/test/regress/sql/partition_pruning.sql]: Makes sense. Added 35 rows of data. Output as expected.

**Comment by ashwinstar** [comment]
Closing this PR based on concerns raised by @chrishajas
